src/ray/gcs/gcs_server/gcs_health_check_manager.cc

```diff
@@ -124,37 +124,47 @@ void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
     return;
   }
 
-  // Reset the context/request/response for the next request.
-  context_.~ClientContext();
-  new (&context_) grpc::ClientContext();
-  response_.Clear();
+  // grpc context and health check response are dedicated to one single async request.
+  auto context = std::make_shared<grpc::ClientContext>();
```

Comment:
Our intent is to sequential-ize grpc call and next health check, but no guarantee there's no overlapping between these two calls, and how passed in client context is used.

Reply:
Q: StartHealthCheck is only called from either start of the manager, or the completion of the last rpc. then how can the context still be used at this place where we destroy it?

Reply:
Yeah good question, I thought it for a while.

First, io context and async grpc are executed in two threads, so we have to order the access to the same variable;
Second, the functionality for grpc call is: schedule io context call X ms later (and access context variable)

There's no guarantee 
(1) second io context call, and reconstruct context
and (2) grpc later access context
have no concurrent access to the same context

it's pretty rare situation, I only met it twice within 1000 times repeated testing.